CONTRIBUTING.md

```diff
@@ -121,3 +122,120 @@ run it before submitting your PR. It runs `gofmt -s` (simplify) and `golint`.
+error and decide if it is a fatal problem or a transient problem that may go away
+when retried. Fatal errors should be logged or recorded in an internal metric to
+provide visibility to users of the Collector. For transient errors come up with a
+retrying strategy and implement it. Typically you will
+want to implement retries with some sort of exponential back-off strategy. For
+connection or sending retries use jitter for back-off intervals to avoid overwhelming
+your destination when network is restored or the destination is recovered.
+[Exponential Backoff](https://github.com/cenkalti/backoff) is a good library that
+provides all this functionality.
+
+### Logging
+
+Log your component startup and shutdown, including successful outcomes (but don't
+overdo it, keep the number of success message to a minimum).
+This can help to understand the context of failures if they occur elsewhere after
+your code is successfully executed.
+
+Use logging carefully for events that can happen frequently to avoid flooding
+the logs. Avoid outputting logs per a received or processed data item since this can
+amount to very large number of log entries (Collector is designed to process
+many thousands of spans and metrics per second). For such high-frequency events
+instead of logging consider adding an internal metric and increment it when
+the event happens.
+
+Make log message human readable and also include data that is needed for easier
+understanding of what happened and in what context.
+
+### Resource Usage
+
+Limit usage of CPU, RAM or other resources that the code can use. Do not write code
+that consumes resources in an uncontrolled manner. For example if you have a queue
+that can contain unprocessed messages always limit the size of the queue unless you
+have other ways to guarantee that the queue will be consumed faster than items are
+added to it.
+
+Performance test the code for both normal use-cases under acceptable load and also for
+abnormal use-cases when the load exceeds acceptable many times. Ensure that
+your code performs predictably under abnormal use. For example if the code
+needs to process received data and cannot keep up with the receiving rate it is
+not acceptable to keep allocating more memory for received data until the Collector
+runs out of memory. Instead have protections for these situations, e.g. when hitting
+resource limits drop the data and record the fact that it was dropped in a metric
+that is exposed to users.
+
+### Graceful Shutdown
+
+Collector does not yet support graceful shutdown but we plan to add it. All components
+must be ready to shutdown gracefully via `Shutdown()` function that all component
+interfaces require. If components contain any temporary data they need to process
+and export it out of the Collector before shutdown is completed. The shutdown process
+will have a maximum allowed duration so put a limit on how long your shutdown
```

Comment:
a follow up from here: we need to change the interface to pass this time limit, it makes more sense then components deciding on their own about their shutdown timeouts.

Reply:
You reading my thoughts :-)
The reason I opted not to touch it is to avoid breaking interfaces. We can do it. However, since we plan moving to OTLP in-memory representation and that will be another breaking change I was thinking to do it only once.
Perhaps not a big deal, it is a trivial change that we can do across entire codebase here and in contrib easily.